pkg/runner/test_runner.go

```diff
@@ -6,282 +6,796 @@
+	targetCtx, _, targetCancel := statectx.WithParent(ctx)
+	defer targetCancel()
+
+	// Set up the pipeline
+	for i, sb := range t.TestStepsBundles {
+		tr.steps = append(tr.steps, &stepState{
+			stepIndex: i,
+			sb:        sb,
+			inCh:      make(chan *target.Target),
+			outCh:     make(chan *target.Target),
+			errCh:     make(chan cerrors.TargetError),
+			ev: storage.NewTestEventEmitter(testevent.Header{
+				JobID:         jobID,
+				RunID:         runID,
+				TestName:      t.Name,
+				TestStepLabel: sb.TestStepLabel,
+			}),
+			tgtDone: make(map[*target.Target]bool),
+			log:     logging.AddField(tr.log, "step", sb.TestStepLabel),
+		})
+		// Step handlers will be started from target handlers as targets reach them.
+	}
+
+	// Set up the targets
+	tr.targets = make(map[string]*targetState)
+	// If we have target state to resume, do it now.
+	if len(resumeState) > 0 {
+		tr.log.Debugf("Attempting to resume from state: %s", string(resumeState))
+		var rs resumeStateStruct
+		if err := json.Unmarshal(resumeState, &rs); err != nil {
+			return nil, fmt.Errorf("invalid resume state: %w", err)
+		}
+		if rs.Version != resumeStateStructVersion {
+			return nil, fmt.Errorf("incompatible resume state version %d (want %d)",
+				rs.Version, resumeStateStructVersion)
+		}
+		if rs.JobID != jobID {
+			return nil, fmt.Errorf("wrong resume state, job id %d (want %d)", rs.JobID, jobID)
+		}
+		tr.targets = rs.Targets
+	}
+	// Initialize remaining fields of the target structures,
+	// build the map and kick off target processing.
+	for _, tgt := range targets {
+		ts := tr.targets[tgt.ID]
+		if ts == nil {
+			ts = &targetState{
+				CurPhase: targetStepPhaseInit,
+			}
+		}
+		ts.tgt = tgt
+		ts.resCh = make(chan error)
+		tr.mu.Lock()
+		tr.targets[tgt.ID] = ts
+		tr.mu.Unlock()
+		tr.targetsWg.Add(1)
+		go func() {
+			tr.targetHandler(targetCtx, stepCtx, ts)
+			tr.targetsWg.Done()
+		}()
+	}
+
+	// Run until no more progress can be made.
+	runErr := tr.runMonitor()
+	if runErr != nil {
+		tr.log.Errorf("monitor returned error: %q, canceling", runErr)
+		stepCancel()
+	}
+
+	// Wait for step runners and readers to exit.
+	if err := tr.waitStepRunners(ctx); err != nil {
+		tr.log.Errorf("step runner error: %q, canceling", err)
+		stepCancel()
+		if runErr == nil {
+			runErr = err
+		}
+	}
+	// There will be no more results, reel in all the target runners (if any).
+	tr.log.Debugf("waiting for target runners to finish")
+	targetCancel()
+	tr.targetsWg.Wait()
+
+	// Examine the resulting state.
+	tr.log.Debugf("leaving, err %v, target states:", runErr)
+	tr.mu.Lock()
+	defer tr.mu.Unlock()
+	resumeOk := (runErr == nil)
+	var inFlightTargets []*targetState
+	for i, tgt := range targets {
+		ts := tr.targets[tgt.ID]
+		stepErr := tr.steps[ts.CurStep].runErr
+		tr.log.Debugf("  %d %s %v", i, ts, stepErr)
+		if ts.CurPhase == targetStepPhaseRun {
+			inFlightTargets = append(inFlightTargets, ts)
+			if stepErr != statectx.ErrPaused {
```

Comment:
If there is even only one target that failed in any step, we are declaring the test not resumable (but continuing the loop after we declare `resumeOk` false)? It should be fine to fail a target in a step, and still be able to resume the test.

Reply:
stepState.runErr aka stepErr is not about target errors, it is

// Runner error, returned from Run() or an error condition detected by the reader.

i.e. it's an error condition that aborts entire run. presence of such errors making run non-resumable makes sense.

Reply:
Well, ok, not sure why I missed that here we were referring to an error of the whole step.